[CodeCompletion] Report type relations when completing inside result builders

ahoppen opened the pull request:
This requires navigating the constraint system solution to retrieve the argument type of the `buildBlock` call. The comments in the code should describe what I’m doing.

rdar://83846531

ahoppen committed:
[Sema] Print whether a type variable allows binding to a hole

Helped me while debugging and might help others as well.

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen committed:
[CodeCompletion] Record fixes while solving result builders for code completion

We record fixes while solving normal expressions for code completion and we should do the same when solving result builders if we are reporting the solutions to completion callbacks.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

xedin commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -3552,18 +3552,80 @@ Type Solution::simplifyTypeForCodeCompletion(Type Ty) const {
+    // it. Look for a ArgumentConversion constraint that allows us to retrieve
+    // the argument type var.
+    for (auto argConstraint : CS.getConstraintGraph().gatherConstraints(
+             typeVar, ConstraintGraph::GatheringKind::AllMentions)) {
```
I think you don't need `AllMentions` in this case or the latter one `EquivalenceClass` should be sufficient because the constraint in question is associated directly with the type variable the search is performed for.

xedin replied:
Actually it might even be enough to just iterate over direct constraints in both cases via `ConstraintGraphNode::getConstraints()`

xedin reviewed (commented):
`solveForCodeCompletion` changes look good! I left a comment that would be good to address before this is merged.

ahoppen committed:
[CodeCompletion] Report type relations when completing inside result builders

This requires navigating the constraint system solution to retrieve the argument type of the `buildBlock` call. The comments in the code should describe what I’m doing well enough.

rdar://83846531

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test
